### PR TITLE
Add std::panic::propagate

### PR DESCRIPTION
**Pull request by sfackler**
See rust-lang/rfcs#1413.

r? @alexcrichton

### CODE REVIEW

**Comment by sfackler** [comment]
Should this be attached to recover's tracking issue or have its own?

File: src/libstd/sys/common/unwind/mod.rs [hunk 1]
```diff
@@ -172,7 +172,7 @@ pub fn panicking() -> bool {
 #[inline(never)]
 #[no_mangle]
 #[allow(private_no_mangle_fns)]
```
Comment (durka) [inline src/libstd/sys/common/unwind/mod.rs]: presumably can remove this allow if the fn is becoming public?

Comment (sfackler) [inline src/libstd/sys/common/unwind/mod.rs]: Possibly - not sure how deeply that lint looks. This function is now public within its module, but not public outside of the crate.

Comment (durka) [inline src/libstd/sys/common/unwind/mod.rs]: You're right, still required: http://is.gd/Z4qyRE

**Comment by brson** [comment]
cc @rust-lang/libs

File: src/libstd/panic.rs [hunk 2]
```diff
@@ -259,3 +261,28 @@ pub fn recover<F: FnOnce() -> R + RecoverSafe, R>(f: F) -> Result<R> {
...
+///
+/// let result = panic::recover(|| {
+///     panic!("oh no!");
+/// });
+///
+/// if let Err(err) = result {
+///     panic::propagate(err);
+/// }
+/// ```
+#[unstable(feature = "panic_propagate", reason = "awaiting feedback", issue = "0")]
```
Comment (huonw) [inline src/libstd/panic.rs]: This should have an actual issue before it lands.

**Comment by sfackler** [comment]
@bors r=aturon

We discussed this during libs triage today and are okay with it.

**Pull request merged**